core/src/main/java/org/apache/carbondata/scan/filter/FilterUtil.java

```diff
@@ -623,9 +655,11 @@ public static DimColumnFilterInfo getFilterListForAllMembersRS(Expression expres
     if (null != dimColumnFilterInfo) {
       for (Integer surrogate : dimColumnFilterInfo.getFilterList()) {
         try {
-          keys[carbonDimension.getKeyOrdinal()] = surrogate;
-          filterValuesList
-              .add(getMaskedKey(rangesForMaskedByte, blockLevelKeyGenerator.generateKey(keys)));
+          if (surrogate <= dimColumnsCardinality[carbonDimension.getKeyOrdinal()]) {
+            keys[carbonDimension.getKeyOrdinal()] = surrogate;
+            filterValuesList
+                .add(getMaskedKey(rangesForMaskedByte, blockLevelKeyGenerator.generateKey(keys)));
+          }
```

Comment:
Please break the loop if it greater since the surrogates are always in sorted order here.